homeassistant/components/hassmpris/media_player.py

```diff
@@ -0,0 +1,674 @@
+    _attr_supported_features = SUPPORTED_MINIMAL
+    _attr_playback_rate: float = 1.0
+    _attr_should_poll: bool = False
+
+    def __init__(
+        self,
+        client: hassmpris_client.AsyncMPRISClient,
+        integration_id: str,
+        player_id: str,
+        initial_state: str | None = None,
+    ) -> None:
+        """
+        Initialize the entity.
+
+        Parameters:
+          client: the client to the remote agent
+          integration_id: unique identifier of the integration
+          player_id: the name / unique identifier of the player
+        """
+        super().__init__()
+        if client is None:
+            raise ValueError("Instantiation of this class requires a client")
+        self.client: hassmpris_client.AsyncMPRISClient | None = client
+        self._client_host = self.client.host
+        self.player_id = player_id
+        self._integration_id = integration_id
+        self._attr_available = True
+        self._metadata: dict[str, Any] = {}
+        if initial_state is not None:
+            self._attr_state = initial_state
+
+    async def set_unavailable(self):
+        """Mark player as unavailable."""
+        _LOGGER.debug("Marking %s as unavailable", self.name)
+        self.client = None
+        self._attr_available = False
+        await self.update_state(STATE_UNKNOWN)
+
+    async def set_available(
+        self,
+        client: hassmpris_client.AsyncMPRISClient,
+    ):
+        """
+        Mark a player as available again.
+
+        Parameters:
+          client: the new client to use to talk to the agent
+        """
+        _LOGGER.debug("Marking %s as available", self.name)
+        self.client = client
+        self._client_host = self.client.host
+        self._attr_available = True
+        if self.hass:
+            await self.async_update_ha_state(True)
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID of this entity."""
+        return self._integration_id + "-" + self.player_id
+
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        assert self.player_id
+        return self.player_id
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return the device information associated with the entity."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self._integration_id)},
+            name="MPRIS agent at %s" % self._client_host,
+            manufacturer="Freedesktop",
+        )
+
+    @staticmethod
+    def config_schema():
+        """Return the discovery schema."""
+        return DISCOVERY_SCHEMA
+
+    async def async_added_to_hass(self) -> None:
+        """Entity has been added to HASS."""
+        _LOGGER.debug("Added to hass: %s", self)
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Entity is about to be removed from HASS."""
+        _LOGGER.debug("Will remove from hass: %s", self)
+        await self.set_unavailable()
+
```

Comment:
```suggestion
```

No need to set unavailable, its being removed

Reply:
Yes we do, because this is where we clear the reference to the `client` object that the entity has.  I also want the logging to remain here, since it has helped me debug object lifecycle issues.

I will change the `set_unavailable()` call to simply remove the `client` reference instead.

Leaving this comment open because I'm very interested in further comments from you.

Reply:
We no longer clear the reference to client or set unavailable.